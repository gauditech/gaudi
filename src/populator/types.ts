type PopulatorModel = {
  model: string;
  repeat: { min: number; max: number } | number;
  data: PopulatorData[];
};

type PopulatorData = {
  field: string;
  type: "constant" | "";
};
