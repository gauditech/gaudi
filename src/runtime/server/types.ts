import { NextFunction, Request, Response } from "express";

export type ServerHttpMethods = "all" | "get" | "post" | "put" | "patch" | "delete";

export type ServerRequestHandler =
  | ((request: Request, response: Response) => void | Promise<void>)
  | ((req: Request, resp: Response, next: NextFunction) => void | Promise<void>);

export type EndpointConfig = {
  path: string;
  method: ServerHttpMethods;
  handlers: ServerRequestHandler[];
};

export type ServerMiddlewareNextFn = (err?: unknown) => void;
