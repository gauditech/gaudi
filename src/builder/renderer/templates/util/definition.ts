import { Definition, PathFragment } from "@src/types/definition";

/*
 * Template helper functions for Gaudi definition
 */

export function modelDbName(modelRefKey: string, definition: Definition) {
  return (definition.models || []).find((model) => model.refKey === modelRefKey)?.dbname;
}

export function fieldDbName(modelRefKey: string, fieldRefKey: string, definition: Definition) {
  return (definition.models || [])
    .find((model) => model.refKey === modelRefKey)
    ?.fields.find((field) => field.refKey === fieldRefKey)?.dbname;
}

export function typeToDbType(type: string) {
  if (type === "serial") {
    return "Int";
  } else if (type === "integer") {
    return "Int";
  } else if (type === "text") {
    return "String";
  } else if (type === "boolean") {
    return "Boolean";
  }

  throw "UNHANDLED_GAUDI_TYPE";
}

export function buildEndpointPath(parts: PathFragment[]): string {
  return parts.reduce((accum, part) => {
    const pathPart =
      part.type === "numeric" || part.type === "text" ? `:${part.varname}` : part.value;
    return `${accum}/${pathPart}`;
  }, "");
}
