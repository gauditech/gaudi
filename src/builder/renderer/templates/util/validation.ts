import { commaLists, oneLine } from "common-tags";

/**
 * Functions for rendering validation schema calls base on Yup (https://github.com/jquense/yup)
 */

/**
 * Render validator object instance
 */
export function renderValidator() {
  return `yup`;
}

// ----- types

/**
 * Render "object" Yup schema
 *
 * https://github.com/jquense/yup#object
 */
export function renderObject<T>(
  fields: Record<string, T>,
  callback: (field: T) => string,
  condition = true
): string {
  // prettier-ignore
  return optionalCall(condition)(commaLists`
    object({
      ${Object.entries<T>(fields).map(([name, field]) => oneLine`"${name}": ${renderValidator()}${callback(field)}`)}
    })
  `);
}

/**
 * Render "mixed" Yup schema
 *
 * https://github.com/jquense/yup#mixed
 */
export function renderMixed(condition = true): string {
  return optionalCall(condition)("mixed()");
}

/**
 * Render "string" Yup schema
 *
 * https://github.com/jquense/yup#string
 */
export function renderString(condition = true): string {
  return optionalCall(condition)("string()");
}

/**
 * Render "number" Yup schema
 *
 * https://github.com/jquense/yup#number
 */
export function renderNumber(condition = true): string {
  return optionalCall(condition)("number()");
}

/**
 * Render "boolean" Yup schema
 *
 * https://github.com/jquense/yup#boolean
 */
export function renderBoolean(condition = true): string {
  return optionalCall(condition)("boolean()");
}

// ----- conditionals

/**
 * Render "required" Yup schema
 *
 * https://github.com/jquense/yup#schemarequiredmessage-string--function-schema
 */
export function renderRequired(condition = true): string {
  return optionalCall(condition)("required()");
}

// ----- utils

/** Returns call string prefixed with "." if condition is true, otherwise it returns an empty string.  */
export function optionalCall(condition: boolean): (call: string) => string {
  return (call) => (condition ? "." + call : "");
}
