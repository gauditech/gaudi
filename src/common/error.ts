import { Interval } from "ohm-js";

export type WithContext<T> = { interval?: Interval } & T;

export class CompilerError extends Error {
  constructor(message: string, context?: WithContext<unknown>) {
    if (!context?.interval) {
      super("Unknown source position!\n" + message);
    } else {
      super(context.interval.getLineAndColumnMessage() + "message");
    }
  }
}
