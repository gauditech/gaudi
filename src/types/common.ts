export type HookCode = HookInline | HookSource;
export type HookInline = { kind: "inline"; inline: string };
export type HookSource = { kind: "source"; target: string; file: string; runtimeName: string };
