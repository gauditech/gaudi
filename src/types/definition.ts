export type Definition = {
  models: ModelDef[];
};

export type ModelDef = {
  refKey: string;
  name: string;
  dbname: string;
  fields: FieldDef[];
  references: ReferenceDef[];
  relations: RelationDef[];
  queries: QueryDef[];
};

export type FieldDef = {
  refKey: string;
  modelRefKey: string;
  name: string;
  dbname: string;
  type: "integer" | "text" | "boolean";
  dbtype: "serial" | "integer" | "text" | "boolean";
  primary: boolean;
  unique: boolean;
  nullable: boolean;
};

export type ReferenceDef = {
  refKey: string;
  modelRefKey: string;
  name: string;
  fieldRefKey: string;
  toModelRefKey: string;
  toModelFieldRefKey: string;
  nullable: boolean;
  unique: boolean;
};

export type RelationDef = {
  refKey: string;
  modelRefKey: string;
  name: string;
  fromModel: string;
  fromModelRefKey: string;
  through: string;
  throughRefKey: string;
  nullable: boolean;
  unique: boolean;
};

type QueryDefPathSelect = {
  name: string;
  retType: string;
  nullable: boolean;
  refKey: string;
};

export type QueryDefPath = {
  refKey: string;
  name: string;
  retType: string;
  retCardinality: "one" | "many"; // should be retCard,...
  nullable: boolean;
  joinType: "inner" | "left";
  alias: string;
  bpAlias: string | null;
  joinPaths: QueryDefPath[];
  select: QueryDefPathSelect[];
};

type QueryDefFilter = {
  type: string;
  lhs: string | number | boolean;
  rhs: string | number | boolean;
};

export type QueryDef = {
  refKey: string;
  name: string;
  retType: string;
  retCardinality: "one" | "many";
  nullable: boolean;
  // unique: boolean;
  joinPaths: QueryDefPath[];
  filters: QueryDefFilter[];
};
