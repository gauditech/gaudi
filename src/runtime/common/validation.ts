import {
  AnySchema,
  BaseSchema,
  BooleanSchema,
  NumberSchema,
  StringSchema,
  ValidationError,
  boolean,
  number,
  object,
  string,
} from "yup";

import { getHook } from "../hooks";

import { assertUnreachable } from "@src/common/utils";
import { BusinessError } from "@src/runtime/server/error";
import {
  FieldsetDef,
  FieldsetFieldDef,
  FieldsetRecordDef,
  HookValidator,
} from "@src/types/definition";

// ----- validation&transformation

export async function validateEndpointFieldset<R = Record<string, unknown>>(
  fieldset: FieldsetDef,
  data: Record<string, unknown>
): Promise<R> {
  try {
    const validationSchema = buildFieldsetValidationSchema(fieldset);
    return await validateRecord(data, validationSchema);
  } catch (err: unknown) {
    if (err instanceof ValidationError) {
      // error should be an instance of Yup's ValidationError
      // https://github.com/jquense/yup#validationerrorerrors-string--arraystring-value-any-path-string

      throw new BusinessError(
        "ERROR_CODE_VALIDATION",
        "Validation error",
        createRecordValidationError(err)
      );
    } else {
      throw err;
    }
  }
}

export type RecordValidationError = Pick<
  ValidationError,
  "value" | "path" | "type" | "errors" | "params" | "inner"
>;

export function createRecordValidationError(error: ValidationError): RecordValidationError {
  return {
    value: error.value,
    path: error.path,
    type: error.type,
    errors: error.errors,
    params: error.params,
    inner: error.inner,
  };
}

export async function validateRecord(record: unknown, schema: AnySchema) {
  return schema.validate(record, {
    abortEarly: false, // report ALL errors, not just the first one
  });
}

// ---------- fieldset validation builder

export function buildFieldsetValidationSchema(fieldset: FieldsetDef): AnySchema {
  if (fieldset.kind !== "record") throw new Error('Root fieldset must be of kind "record".');

  return buildObjectValidationSchema(fieldset);
}

function processFields(name: string, field: FieldsetDef): AnySchema {
  if (field.kind === "field") {
    return buildFieldValidationSchema(name, field);
  } else {
    return buildObjectValidationSchema(field);
  }
}

function buildObjectValidationSchema(field: FieldsetRecordDef): AnySchema {
  const fieldSchemas = Object.fromEntries(
    Object.entries(field.record).map(([name, value]) => [name, processFields(name, value)])
  );

  if (!field.nullable) object(fieldSchemas).required();

  return object(fieldSchemas).optional();
}

function buildFieldValidationSchema(name: string, field: FieldsetFieldDef): AnySchema {
  if (field.type === "text") {
    // start with nullable because it's the only way to
    let s = string();

    if (field.nullable) {
      // TODO: yup's types don't allow expanding return type to `string | undefined | null`
      s = s.nullable() as StringSchema;
    }
    if (field.required) {
      // everything except `undefined`
      s = s.defined();
    }

    field.validators.forEach((v) => {
      if (v.name === "minLength") {
        s = s.min(v.args[0].value);
      } else if (v.name === "maxLength") {
        s = s.max(v.args[0].value);
      } else if (v.name === "isEmail") {
        s = s.email();
      } else if (v.name === "isTextEqual") {
        // TODO: s.equals returns BaseSchema which doesn't fit StringSchema
        s = s.equals<string>([v.args[0].value]) as StringSchema;
      } else if (v.name === "hook") {
        s = buildHookSchema(name, v, s);
      }
    });

    return s;
  } else if (field.type === "integer") {
    let s = number();

    if (field.nullable) {
      // TODO: yup's types don't allow expanding return type to `number | undefined | null`
      s = s.nullable() as NumberSchema;
    }
    if (field.required) {
      // everything except `undefined`
      s = s.defined();
    }

    field.validators.forEach((v) => {
      if (v.name === "min") {
        s = s.min(v.args[0].value);
      } else if (v.name === "max") {
        s = s.max(v.args[0].value);
      } else if (v.name === "isIntEqual") {
        // TODO: s.equals returns BaseSchema which doesn't fit NumberSchema
        s = s.equals([v.args[0].value]) as NumberSchema;
      } else if (v.name === "hook") {
        s = buildHookSchema(name, v, s);
      }
    });

    return s;
  } else if (field.type === "boolean") {
    let s = boolean();

    // NOTE: boolean schema cannot be "nullable"

    if (field.required) {
      // everything except `undefined`
      s = s.defined();
    }

    field.validators.forEach((v) => {
      if (v.name === "isBoolEqual") {
        // TODO: s.equals returns BaseSchema which doesn't fit BooleanSchema
        s = s.equals([v.args[0].value]) as BooleanSchema;
      } else if (v.name === "hook") {
        s = buildHookSchema(name, v, s);
      }
    });

    return s;
  } else {
    assertUnreachable(field.type);
  }
}

function buildHookSchema<S extends BaseSchema>(
  name: string,
  validator: HookValidator,
  schema: S
): S {
  let testFn: (value: unknown) => boolean;
  const code = validator.code;
  const args = validator.args;
  const hasArg = args.find((arg) => arg.name === name) != null;

  if (code.kind === "inline") {
    testFn = (value) => {
      return eval(hasArg ? `const ${name} = ${JSON.stringify(value)};${code.inline}` : code.inline);
    };
  } else {
    testFn = (value) => {
      const hook = getHook(code.file, code.target);
      if (hasArg) {
        return hook(value);
      } else {
        return hook();
      }
    };
  }

  return schema.test(testFn);
}
