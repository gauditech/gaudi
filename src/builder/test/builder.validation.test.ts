import { buildFieldsetValidationSchema } from "@src/runtime/common/validation";
import { FieldsetDef } from "@src/types/definition";

describe("builder", () => {
  describe("validation", () => {
    it("renders fieldset validation schema correctly", async () => {
      const data: FieldsetDef = {
        kind: "record",
        nullable: true, // required record
        record: {
          optional: {
            kind: "field",
            nullable: true, // optional field
            type: "integer", // integer field
          },
          required: {
            kind: "field",
            nullable: false, // required field
            type: "text", // text field
          },
          something: {
            kind: "field",
            nullable: false,
            type: "boolean", // boolean field
          },
          subrecord: {
            // subrecord
            kind: "record",
            nullable: false, // optional record
            record: {
              prop: {
                kind: "field",
                nullable: false,
                type: "text",
              },
            },
          },
        },
      };

      expect(buildFieldsetValidationSchema(data)).toMatchSnapshot();
    });
  });
});
