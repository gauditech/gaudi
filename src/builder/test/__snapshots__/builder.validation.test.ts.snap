// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`builder validation renders fieldset validation schema correctly 1`] = `
"yup
.object({
  "optional": yup.number(),
  "required": yup.string() .required(),
  "something": yup.boolean() .required(),
  "subrecord": yup.object({ "prop": yup.string() .required() })
})"
`;
